### PR TITLE
Add an optional legend to the heatmap

### LINKED ISSUE #119: Add legend to heatmap visualization
There's no legend available for the heatmap visualization. It would be nice if we could add the option to enable such a legend.

### PR DESCRIPTION
**Pull request by bmesuere**
The heatmap maps values onto a continuous color scale, but nothing said which color stood for which value. This adds an opt-in legend: a color bar underneath the grid, labelled with the values it covers.

The bar is drawn from the same discrete palette the grid uses (`minColor`, `maxColor` and `colorBuckets`), not from a hardcoded scale, so it also shows how coarse the scale is when few color buckets are configured. That palette is now computed once in `Preprocessor.computeColorPalette` and shared by the grid and the legend.

The legend takes its height out of the configured `height`, so a heatmap with a legend still measures exactly `width` by `height`. It is rendered as an SVG sibling of the canvas, kept in sync by `resize()`, and left untouched by `cluster()` and `reset()`. `toSVG()` includes it as well, so an exported heatmap keeps the scale that explains its colors. Both are built from one list of shapes computed by `computeLegendShapes`, so the live legend and the exported one cannot drift apart.

## New public API

All of it defaults to off, so existing heatmaps render exactly as before.

| Setting | Default | What it does |
| --- | --- | --- |
| `enableLegend` | `false` | Renders the legend underneath the grid, and in `toSVG()` output |
| `legend.title` | `""` | Title above the color bar. Nothing is rendered when empty |
| `legend.titleFontSize` | `14` | Title size in pixels |
| `legend.width` | `240` | Width of the color bar in pixels, shrunk when less space is available |
| `legend.height` | `12` | Height of the color bar in pixels |
| `legend.labelFontSize` | `12` | Tick label size in pixels |
| `legend.ticks` | `5` | Requested number of ticks, rounded to nice values |
| `legend.tickFormat` | `value => value.toFixed(2)` | Turns a value from `[0, 1]` into a tick label |
| `legend.padding` | `{ top: 8, right: 0, bottom: 0, left: 0 }` | Padding around the legend area |

`HeatmapLegendSettings` is exported alongside `HeatmapSettings`. A partial `legend` object keeps the defaults for the settings it does not mention, `padding` included.

### Heights that cannot fit the legend

When the configured `height` is smaller than the legend needs, the grid is clamped to a single pixel rather than the construction being refused. Heatmaps are routinely resized to whatever a container happens to measure, and a container that is briefly shorter than the legend should not throw and take the page down with it. The result is a visualization slightly taller than requested, which is the lesser problem, and no invalid height ever reaches the canvas.

## Manual testing

Run `npm run build`, serve the repository and open **`examples/heatmap-legend.html`**.

- The legend is visible under the grid: the title "Similarity", the blue color bar, and ticks from 0% to 100%.
- **Toggle legend** rebuilds the heatmap without a legend. The grid grows to fill the space the legend used.
- **Cluster** reorders the rows and columns. The legend stays put.
- **Resize** redraws at 1000 x 400. The legend stays under the grid and the whole thing is 400 pixels tall.
- **Export SVG** downloads `heatmap.svg`. Opening it shows the grid, the labels and the same legend underneath.

The other two heatmap examples are unchanged and must still look the same, since the legend is off by default.

<details>
<summary>What was verified</summary>

- `npm run lint`, `npm run typecheck` and `npm test` are all green. 77 tests, up from 65.
- The four existing heatmap image snapshots still match without being regenerated, which is the check that the default really is off.
- Ten new specs in `src/visualizations/heatmap/__tests__/Heatmap.spec.ts`: no legend by default, legend present when enabled, defaults kept for a partial `legend` object and for a partial `legend.padding`, bar colors equal to the grid palette, legend height subtracted from the grid height, a grid that stays valid when the height cannot fit the legend, the legend surviving `resize()`, `cluster()` and `reset()`, and `toSVG()` carrying the legend only when it is enabled.
- Each of the three review findings has a spec that fails on the previous commit with exactly the reported symptom (legend height `NaN`, canvas height `-20`, no legend group in the export) and passes on this one.
- `examples/heatmap-legend.html` loaded in puppeteer against a built `dist`: no console errors, canvas 540px plus legend 60px equals the configured 600px, after `resize(1000, 400)` canvas 340px plus legend 60px equals 400px, and toggling the legend off returns the canvas to the full 600px.
- The downloaded `heatmap.svg` was rendered in the browser: 1163 by 679 pixels, grid and labels unchanged, legend underneath them with the same title and ticks.
- In the browser, a heatmap built at a height of 20 pixels with `legend: { padding: { top: 4 } }` now gives a 1 pixel canvas and a 36 pixel legend, both on construction and after `resize(400, 20)`. Before, that was a 150 pixel canvas and a `NaN` legend.

</details>

Closes #119

### CODE REVIEW

File: src/visualizations/heatmap/Heatmap.ts [hunk 1]
```diff
@@ -130,14 +143,23 @@ export default class Heatmap {
...
         this.redraw();
     }
 
     private fillOptions(options: any = undefined): HeatmapSettings {
         const output = new HeatmapSettings();
-        return Object.assign(output, options);
+        Object.assign(output, options);
+        // Object.assign is shallow, so a legend object that only mentions some of the settings would otherwise wipe
+        // out the defaults for all the others.
+        output.legend = Object.assign(new HeatmapLegendSettings(), options?.legend);
```
Comment (bmesuere) [inline src/visualizations/heatmap/Heatmap.ts]: 🤖 [P2] Merge the nested padding defaults as well. This merge preserves omitted top-level legend fields, but a normal JavaScript option such as `legend: { padding: { top: 5 } }` replaces the entire padding object, leaving `right`, `bottom`, and `left` undefined. I reproduced the resulting canvas and SVG heights as `NaN`, so the legend does not render. Please deep-merge `padding` with its defaults (and add a partial-padding regression test), matching the advertised partial-settings behavior.

File: src/visualizations/heatmap/Heatmap.ts [hunk 2]
```diff
@@ -288,21 +310,154 @@ export default class Heatmap {
...
+
         this.zoomed(this.lastZoomStatus);
     }
 
+    /**
+     * Height that's available to the heatmap grid itself. The legend is rendered underneath the grid and takes away
+     * part of the configured height, so that the visualization as a whole keeps the dimensions that were requested.
+     */
+    private get gridHeight(): number {
+        return this.settings.height - this.legendHeight();
```
Comment (bmesuere) [inline src/visualizations/heatmap/Heatmap.ts]: 🤖 [P2] Guard against the legend consuming the entire configured height. With the default 40px legend, constructing or resizing a heatmap to 20px makes `gridHeight` equal `-20`; the canvas receives `height="-20"`, its CSS height is rejected, and the browser falls back to the canvas default height of 150px. The component therefore becomes much taller than requested and its viewport calculations are invalid. Clamp to a valid grid height or reject dimensions smaller than the legend, and cover the compact-resize case.

File: src/visualizations/heatmap/Heatmap.ts [hunk 3]
```diff
@@ -130,14 +143,23 @@ export default class Heatmap {
         // @ts-expect-error
         this.visElement.call(zoom);
 
+        if (this.settings.enableLegend) {
+            this.legendElement = d3.select(this.element).append("svg");
```
Comment (bmesuere) [inline src/visualizations/heatmap/Heatmap.ts:164]: 🤖 [P2] Include the enabled legend in the public `toSVG()` export. The legend only exists as this live DOM sibling, while `toSVG()` still serializes the cells and axis labels alone. Consumers enabling the legend to explain the color scale therefore lose that scale in downloaded/exported heatmaps. Please render the same palette/title/ticks into the returned SVG (or explicitly expose and document a separate export path if omission is intentional) and add an export regression test.

**Review by bmesuere — COMMENTED** [review]
🤖 Found three edge cases in the new legend API at `d803d43`. Typecheck, lint, build, and all 60 non-snapshot tests pass; the 11 local snapshot failures are the existing Node 25/Puppeteer `Uint8Array` versus `Buffer` matcher incompatibility.

File: src/visualizations/heatmap/Heatmap.ts (hunk 1 above)
Comment (bmesuere) [reply src/visualizations/heatmap/Heatmap.ts]: Fixed in aaf510c. `fillOptions` now runs the legend through a `fillGroup` helper that merges the nested padding with its defaults first, the same shape as the one in #212:

```ts
private fillGroup<T extends { padding: VisualizationPadding }>(defaults: T, options: any): T {
    const padding = Object.assign({}, defaults.padding, options?.padding);
    return Object.assign(defaults, options, { padding });
}
```

Regression test: "should fall back to the default padding for the parts of it that are not given" builds a heatmap with `legend: { padding: { top: 4 } }` and asserts a legend height of 36 and a canvas that fills the rest of the configured 400. It fails on the previous commit with `expected NaN to be 36`. Checked in the browser too: 1 pixel canvas, 36 pixel legend, no NaN.

File: src/visualizations/heatmap/Heatmap.ts (hunk 2 above)
Comment (bmesuere) [reply src/visualizations/heatmap/Heatmap.ts]: Fixed in aaf510c. `gridHeight` is now `Math.max(MINIMUM_GRID_HEIGHT, height - legendHeight())`, so nothing below one pixel can reach the canvas from either the constructor or `resize()`.

I picked clamping over refusing the size. Heatmaps get resized to whatever a container happens to measure, so a container that is briefly shorter than the legend would throw during an ordinary window resize and take the page with it. The trade off is a visualization a few pixels taller than requested in that degenerate case, which I noted in the PR body.

Regression test: "should keep a usable grid when the configured height is smaller than the legend" constructs at 800 by 20 with the legend on and asserts a positive canvas height, then repeats after `resize(800, 20)`. It fails on the previous commit with `expected -20 to be greater than 0`. In Chrome the canvas is now 1 pixel tall instead of falling back to 150.

File: src/visualizations/heatmap/Heatmap.ts (hunk 3 above)
Comment (bmesuere) [reply src/visualizations/heatmap/Heatmap.ts:164]: Agreed, and fixed in aaf510c. The export is exactly the case where a colour scale needs explaining, so leaving it out was the wrong call.

To keep the two from drifting, the geometry is computed once: `computeLegendShapes(barWidth)` returns a list of rects, lines and texts positioned relative to the legend origin. `renderLegend` turns those into d3 elements next to the canvas, `legendShapesToSVG` turns the same list into markup for `toSVG`. Only the available width differs, which both take through `legendBarWidth`. The legend is placed under the axis labels and grows the exported image rather than overlapping it, and is only emitted when `enableLegend` is on.

Regression tests: "should export the legend as part of the SVG when it is enabled" asserts the legend group, the title and the last tick are in the string, and "should not export a legend when it is disabled" asserts the group is absent. The first fails on the previous commit. The example page has an Export SVG button now; the downloaded file renders as the grid, the labels and the legend underneath, matching the screen.
